### PR TITLE
Compiler Fixes

### PR DESCRIPTION
**Pull request by trshaffer**
This PR fixes a bunch of compiler warnings (mostly due to un-checked return codes). Strict build works for me under Nix (for #2326  and #2327). There are quite a few changes, so I was just going for quick error checks (e.g. crash on unexpected failure).

### CODE REVIEW

**Comment by dthain** [comment]
@trshaffer let's talk about this a little more.

What's the compiler flag that's forcing these errors?
Is it finding real errors, or misleading ones?
And do we want all of the problems fixed in this way?

From a quick look, it seems to me that some of these are bugs, and some are not.  And some of those that are bugs should be addressed by a proper error return instead of an assertion failure.

**Comment by trshaffer** [comment]
Aside from `gettimeofday()` (discussion in #2327), these are warnings due to not checking return codes. A number of libc functions are marked `__attribute__((warn_unused_result))` which requires their return values to be consumed. Thatt gives a few options:
- Ignore it. We won't be able to compile with `--strict` on some newer compilers.
- Disable warnings. The `-Wno-unused-result` flag would disable this. However, I think we generally do want to get these warnings, as they strongly suggest the presence of a bug.
- Disable warnings at these locations. GCC can do localized changes to warnings (via `#pragma GCC diagnostic`). However, this is a GCC-specific feature, and again I think these warnings ought to be fixed. There are also some incantations (GCC's `__attribute__ ((unused))`, or casting to `void`) that can suppress this warning for a single location.
- Fixing the code to check return values.

I think in almost all cases (some of the error messages using `write()` being the exception) that these are indeed bugs in our code. However since there are quite a few of them spread across many different programs, fixing them would require a small campaign. The diff for this PR gives a listing of all of them and captures the returns in variables, but doesn't really do the right thing in most cases (just crashes). This complete listing however is useful to triage warnings and decide what needs doing. I'd propose we assign`.c` files to whomever most recently worked on the code (or "owns" that part of the codebase) and add patches to emit useful warnings, abort, etc. to this PR. @dthain thoughts?

**Comment by dthain** [comment]
Ok, I would like to split the instances into three cases:

1 - Cases where we should detect failure and return it up the chain.   For example `batch_job_blue_waters_remove` should simply be `return rc==0;` 

2 - Cases where we should just ignore the return value, typically b/c the operation is informative, or will be retried in some other way, or we are in the middle of a failure.  I suggest that what we do here is introduce a macro `IGNORE` and define that to be the gcc magic, so that we don't pollute our code everywhere with gcc-isms.

3 - Cases where we should abort immediately.  In that case, we should be using the cctools `fatal` rather than `assert` to have consistent behavior with other kinds of failures.

I'm going to run through the PR and comment on each of these as DETECT, IGNORE, or FATAL as appropriate...

File: batch_job/src/batch_job_blue_waters.c [hunk 1]
```diff
@@ -255,7 +256,8 @@ static int batch_job_cluster_remove (struct batch_queue *q, batch_job_id_t jobid
 	info->exit_signal = 1;
 
 	char *command = string_format("%s %" PRIbjid, cluster_remove_cmd, jobid);
-	system(command);
+	int rc = system(command);
```
Comment (dthain) [inline batch_job/src/batch_job_blue_waters.c:259]: DETECT

File: batch_job/src/batch_job_cluster.c [hunk 2]
```diff
@@ -68,7 +69,8 @@ static int setup_batch_wrapper(struct batch_queue *q, const char *sysname )
 	fchmod(fileno(file), 0755);
 
 	char path[PATH_MAX];
-	getcwd(path, PATH_MAX);
+	char *rc = getcwd(path, PATH_MAX);
```
Comment (dthain) [inline batch_job/src/batch_job_cluster.c:72]: DETECT

File: batch_job/src/batch_job_cluster.c [hunk 3]
```diff
@@ -338,7 +340,8 @@ static int batch_job_cluster_remove (struct batch_queue *q, batch_job_id_t jobid
 	info->exit_signal = 1;
 
 	char *command = string_format("%s %" PRIbjid, cluster_remove_cmd, jobid);
-	system(command);
+	int rc = system(command);
```
Comment (dthain) [inline batch_job/src/batch_job_cluster.c:343]: DETECT

File: batch_job/src/batch_job_condor.c [hunk 4]
```diff
@@ -299,7 +300,8 @@ static batch_job_id_t batch_job_condor_wait (struct batch_queue * q, struct batc
 
 					info->finished = current;
 
-					fgets(line, sizeof(line), logfile);
+					char *rc = fgets(line, sizeof(line), logfile);
```
Comment (dthain) [inline batch_job/src/batch_job_condor.c:303]: DETECT

File: batch_job/src/batch_job_internal.h [hunk 5]
```diff
@@ -65,7 +66,7 @@ struct batch_queue {
 #define batch_queue_stub_option_update(name)  static void batch_queue_##name##_option_update (struct batch_queue *Q, const char *what, const char *value) { return; }
 
 #define batch_fs_stub_chdir(name)  static int batch_fs_##name##_chdir (struct batch_queue *Q, const char *path) { return chdir(path); }
-#define batch_fs_stub_getcwd(name)  static int batch_fs_##name##_getcwd (struct batch_queue *Q, char *buf, size_t size) { getcwd(buf, size); return 0; }
+#define batch_fs_stub_getcwd(name)  static int batch_fs_##name##_getcwd (struct batch_queue *Q, char *buf, size_t size) { char *rc = getcwd(buf, size); assert(rc != NULL); return 0; }
```
Comment (dthain) [inline batch_job/src/batch_job_internal.h:69]: DETECT

File: batch_job/src/batch_job_k8s.c [hunk 6]
```diff
@@ -573,7 +574,8 @@ static batch_job_id_t batch_job_k8s_wait (struct batch_queue * q,
 			if(!cmd_fp) {
 				return -1;
 			}
-			fgets(log_tail_content, sizeof(log_tail_content)-1, cmd_fp);
+			char *rc = fgets(log_tail_content, sizeof(log_tail_content)-1, cmd_fp);
```
Comment (dthain) [inline batch_job/src/batch_job_k8s.c:577]: DETECT

File: batch_job/src/batch_job_k8s.c [hunk 7]
```diff
@@ -679,7 +681,8 @@ static int batch_queue_k8s_free(struct batch_queue *q)
 {
 	char *cmd_rm_tmp_files = string_format("rm %s-*.json %s %s", 
 			mf_uuid->str, k8s_script_file_name, kubectl_failed_log);
-	system(cmd_rm_tmp_files);
+	int rc = system(cmd_rm_tmp_files);
```
Comment (dthain) [inline batch_job/src/batch_job_k8s.c:684]: DETECT

File: batch_job/src/work_queue_factory.c [hunk 8]
```diff
@@ -124,7 +125,8 @@ invoke, so we construct a string and emit it with a low-level write.
 static void handle_abort( int sig )
 {
 	const char *msg = "received abort signal, shutting down workers...\n";
-	write(1,msg,strlen(msg));
+	ssize_t rc = write(1,msg,strlen(msg));
```
Comment (dthain) [inline batch_job/src/work_queue_factory.c:128]: IGNORE

File: batch_job/src/work_queue_factory.c [hunk 9]
```diff
@@ -1422,7 +1424,8 @@ int main(int argc, char *argv[])
 
 	if(password_file) {
 		cmd = string_format("cp %s %s/pwfile",password_file,scratch_dir);
-		system(cmd);
+		int rc = system(cmd);
```
Comment (dthain) [inline batch_job/src/work_queue_factory.c:1427]: DETECT or FATAL

File: chirp/src/chirp_audit.c [hunk 10]
```diff
@@ -118,7 +119,8 @@ struct hash_table *chirp_audit(const char *path)
 	   anyone else.
 	 */
 
-	nice(10);
```
Comment (dthain) [inline chirp/src/chirp_audit.c:121]: IGNORE

File: chirp/src/chirp_matrix.c [hunk 11]
```diff
@@ -98,7 +99,8 @@ struct chirp_matrix *chirp_matrix_create(const char *host, const char *path, int
 	for(i = 0; (int) i < nhosts; i++) {
 		if(!fgets(line, sizeof(line), file)) {
 			rewind(file);
-			fgets(line, sizeof(line), file);
+			char *rc = fgets(line, sizeof(line), file);
+			assert(rc != NULL);
```
Comment (dthain) [inline chirp/src/chirp_matrix.c:103]: FATAL

File: chirp/src/chirp_server.c [hunk 12]
```diff
@@ -2132,7 +2133,8 @@ int main(int argc, char *argv[])
 		debug(D_CHIRP, "transient directory: `%s'", chirp_transient_path);
 	}
 
-	chdir("/"); /* no more relative path access from this point on */
+	int rc = chdir("/"); /* no more relative path access from this point on */
+	assert(rc == 0);
```
Comment (dthain) [inline chirp/src/chirp_server.c:2137]: FATAL

File: chirp/src/chirp_stats.c [hunk 13]
```diff
@@ -116,7 +117,8 @@ void chirp_stats_report(int pipefd, const char *addr, const char *subject, int i
 
 	if(time(0) - child_report_time > interval) {
 		snprintf(line, PIPE_BUF, "stats %s %s %" PRId64 " %" PRId64 " %" PRId64 "\n", addr, subject, child_ops, child_bytes_read, child_bytes_written);
-		write(pipefd, line, strlen(line));
+		ssize_t rc = write(pipefd, line, strlen(line));
+		assert(rc != -1);
```
Comment (dthain) [inline chirp/src/chirp_stats.c:121]: IGNORE

File: chirp/src/chirp_tool.c [hunk 14]
```diff
@@ -181,8 +181,10 @@ static INT64_T do_lcd(int argc, char **argv)
 {
 	char full_path[CHIRP_PATH_MAX];
 	complete_local_path(argv[1], full_path);
+	char *rc;
 	if(chdir(full_path) == 0) {
-		getcwd(current_local_dir, CHIRP_PATH_MAX);
+		rc = getcwd(current_local_dir, CHIRP_PATH_MAX);
+		assert(rc != NULL);
```
Comment (dthain) [inline chirp/src/chirp_tool.c:187]: DETECT

File: chirp/src/chirp_tool.c [hunk 15]
```diff
@@ -1013,12 +1018,15 @@ static INT64_T do_xattr_list(int argc, char **argv)
...
 	char data[65536];
 	INT64_T size;
+	ssize_t rc;
 
 	if((size = chirp_reli_listxattr(current_host, full_path, data, sizeof(data), stoptime)) > 0) {
 		char *current;
 		for(current = data; *current; current = current + strlen(current) + 1) {
-			write(STDOUT_FILENO, current, strlen(current));
-			write(STDOUT_FILENO, "\n", 1);
+			rc = write(STDOUT_FILENO, current, strlen(current));
```
Comment (dthain) [inline chirp/src/chirp_tool.c:1026]: IGNORE

File: dttools/src/bitmap.c [hunk 16]
```diff
@@ -201,17 +202,21 @@ struct bitmap *bitmap_load_raw(const char *path)
...
 	int width, height;
 	struct bitmap *m;
+	size_t rc;
 	file = fopen(path, "rb");
 	if(!file)
 		return 0;
-	fread(&width, 1, sizeof(width), file);
-	fread(&height, 1, sizeof(height), file);
+	rc = fread(&width, 1, sizeof(width), file);
+	assert(rc == sizeof(width));
```
Comment (dthain) [inline dttools/src/bitmap.c:210]: FATAL

File: dttools/src/bitmap.c [hunk 17]
```diff
@@ -325,10 +330,12 @@ struct bitmap *bitmap_load_bmp(const char *path)
 	struct bitmap *m;
 	struct bmp_header header;
 	int i;
+	size_t rc;
 	file = fopen(path, "rb");
 	if(!file)
 		return 0;
-	fread(&header, 1, sizeof(header), file);
+	rc = fread(&header, 1, sizeof(header), file);
```
Comment (dthain) [inline dttools/src/bitmap.c:337]: FATAL

File: dttools/src/disk_alloc.c [hunk 18]
```diff
@@ -185,7 +186,8 @@ int disk_alloc_delete(char *loc) {
 	FILE *loop_find;
 	losetup_args = string_format("losetup -j %s", device_loc);
 	loop_find = popen(losetup_args, "r");
-	fscanf(loop_find, "%s %s %s", loop_dev, loop_info, loop_mount);
+	int rc = fscanf(loop_find, "%s %s %s", loop_dev, loop_info, loop_mount);
```
Comment (dthain) [inline dttools/src/disk_alloc.c:189]: DETECT

File: dttools/src/gpu_info.c [hunk 19]
```diff
@@ -16,15 +16,16 @@ See the file COPYING for details.
...
 
 
 #define GPU_AUTODETECT "cctools_gpu_autodetect"
 
 int gpu_info_get()
 {
 	int pipefd[2];
-	pipe(pipefd);
+	int rc = pipe(pipefd);
+	assert(rc == 0);
```
Comment (dthain) [inline dttools/src/gpu_info.c:28]: DETECT

File: dttools/src/host_memory_info.c [hunk 20]
```diff
@@ -60,7 +61,8 @@ int host_memory_usage_get(UINT64_T * rssp, UINT64_T * totalp)
 	if(!file)
 		return 0;
 
-	fscanf(file, "%lu %lu %lu %lu %lu %lu %lu", &total, &rss, &shared, &text, &libs, &data, &dirty);
+	int rc = fscanf(file, "%lu %lu %lu %lu %lu %lu %lu", &total, &rss, &shared, &text, &libs, &data, &dirty);
```
Comment (dthain) [inline dttools/src/host_memory_info.c:64]: DETECT

File: dttools/src/libforce_halt_enospc.c [hunk 21]
```diff
@@ -70,13 +72,19 @@ ssize_t write(int fd, const void *buf, size_t count) {
 		int fd;
 		char *filename = getenv("CCTOOLS_DISK_ALLOC");
 		if(!filename) {
-			original_write(STDERR_FILENO, "WRITE ERROR: could not set flag to alert resource management system that loop device is full.\n", 94);
-			original_write(STDERR_FILENO, "WRITE ERROR: device capacity reached.\n", 39);
+			rc = original_write(STDERR_FILENO, "WRITE ERROR: could not set flag to alert resource management system that loop device is full.\n", 94);
```
Comment (dthain) [inline dttools/src/libforce_halt_enospc.c:75]: IGNORE

Comment (dthain) [reply dttools/src/libforce_halt_enospc.c:75]: x4

File: work_queue/src/work_queue_worker.c [hunk 22]
```diff
@@ -2933,10 +2938,13 @@ int main(int argc, char *argv[])
 		string_nformat(rm_container_cmd, sizeof(rm_container_cmd), "docker rm %s", container_name);
 
 		if(container_mode == CONTAINER_MODE_DOCKER_PRESERVE) {
+			int rc;
```
Comment (dthain) [inline work_queue/src/work_queue_worker.c:2941]: IGNORE

File: work_queue/src/work_queue_worker.c [hunk 23]
```diff
@@ -2784,7 +2785,8 @@ int main(int argc, char *argv[])
 	}
 
 	// change to workspace
-	chdir(workspace);
+	rc = chdir(workspace);
```
Comment (dthain) [inline work_queue/src/work_queue_worker.c:2788]: FATAL

File: work_queue/src/work_queue_worker.c [hunk 24]
```diff
@@ -2827,22 +2829,25 @@ int main(int argc, char *argv[])
 	if(container_mode == CONTAINER_MODE_DOCKER && load_from_tar == 1) {
 		char load_cmd[1024];
 		string_nformat(load_cmd, sizeof(load_cmd), "docker load < %s", tar_fn);
-		system(load_cmd);
+		int rc = system(load_cmd);
```
Comment (dthain) [inline work_queue/src/work_queue_worker.c:2832]: FATAL

File: work_queue/src/work_queue_worker.c [hunk 25]
```diff
@@ -2827,22 +2829,25 @@ int main(int argc, char *argv[])
...
+		assert(rc == 0);
 	}
 
 	if(container_mode == CONTAINER_MODE_DOCKER_PRESERVE) {
 		if (load_from_tar == 1) {
 			char load_cmd[1024];
 			string_nformat(load_cmd, sizeof(load_cmd), "docker load < %s", tar_fn);
-			system(load_cmd);
+			int rc = system(load_cmd);
```
Comment (dthain) [inline work_queue/src/work_queue_worker.c:2840]: FATAL

File: work_queue/src/work_queue_worker.c [hunk 26]
```diff
@@ -2827,22 +2829,25 @@ int main(int argc, char *argv[])
...
+			assert(rc == 0);
 		}
 
 		string_nformat(container_name, sizeof(container_name), "worker-%d-%d", (int) getuid(), (int) getpid());
 		char container_mnt_point[1024];
 		char start_container_cmd[1024];
 		string_nformat(container_mnt_point, sizeof(container_mnt_point), "%s:%s", workspace, DOCKER_WORK_DIR);
 		string_nformat(start_container_cmd, sizeof(start_container_cmd), "docker run -i -d --name=\"%s\" -v %s -w %s %s", container_name, container_mnt_point, DOCKER_WORK_DIR, img_name);
-		system(start_container_cmd);
+		int rc = system(start_container_cmd);
```
Comment (dthain) [inline work_queue/src/work_queue_worker.c:2849]: FATAL

File: work_queue/src/work_queue_test.c [hunk 27]
```diff
@@ -34,7 +35,8 @@ int submit_tasks(struct work_queue *q, int input_size, int run_time, int output_
 
 	sprintf(input_file, "input.%d",ntasks);
 	sprintf(gen_input_cmd, "dd if=/dev/zero of=%s bs=1048576 count=%d",input_file,input_size);
-	system(gen_input_cmd);
+	int rc = system(gen_input_cmd);
```
Comment (dthain) [inline work_queue/src/work_queue_test.c:38]: FATAL

File: work_queue/src/work_queue.c [hunk 28]
```diff
@@ -4979,7 +4981,8 @@ struct work_queue *work_queue_create(int port)
 		link_address_local(q->master_link, address, &q->port);
 	}
 
-	getcwd(q->workingdir,PATH_MAX);
+	rc = getcwd(q->workingdir,PATH_MAX);
```
Comment (dthain) [inline work_queue/src/work_queue.c:4984]: FATAL

File: work_queue/src/work_queue.c [hunk 29]
```diff
@@ -1801,7 +1801,8 @@ static work_queue_result_code_t get_update( struct work_queue *q, struct work_qu
 
 	lseek(fd,offset,SEEK_SET);
 	link_stream_to_fd(w->link,fd,length,stoptime);
-	ftruncate(fd,offset+length);
+	int rc = ftruncate(fd,offset+length);
```
Comment (dthain) [inline work_queue/src/work_queue.c:1804]: FATAL

File: resource_monitor/src/rmonitor_helper_comm.c [hunk 30]
```diff
@@ -229,8 +230,10 @@ int rmonitor_helper_init(char *lib_default_path, int *fd, int stop_short_running
 	char *helper_path = rmonitor_helper_locate(lib_default_path);
 	char helper_absolute[PATH_MAX + 1];
 	char *rmonitor_port;
+	char *rc;
 
-	realpath(helper_path, helper_absolute);
+	rc = realpath(helper_path, helper_absolute);
```
Comment (dthain) [inline resource_monitor/src/rmonitor_helper_comm.c:235]: FATAL

File: resource_monitor/src/resource_monitor.c [hunk 31]
```diff
@@ -1684,7 +1685,8 @@ void write_helper_lib(void)
         return;
 
     n = sizeof(lib_helper_data);
-    write(flib, lib_helper_data, n);
+    ssize_t rc = write(flib, lib_helper_data, n);
```
Comment (dthain) [inline resource_monitor/src/resource_monitor.c:1688]: FATAL

File: parrot/src/pfs_table.cc [hunk 32]
```diff
@@ -102,7 +102,8 @@ pfs_table::pfs_table()
 	if(pfs_initial_working_directory) {
 		strcpy(working_dir,pfs_initial_working_directory);
 	} else {
-		::getcwd(working_dir,sizeof(working_dir));
+		char *rc = ::getcwd(working_dir,sizeof(working_dir));
```
Comment (dthain) [inline parrot/src/pfs_table.cc:105]: FATAL

File: parrot/src/pfs_main.cc [hunk 33]
```diff
@@ -966,9 +967,9 @@ int main( int argc, char *argv[] )
 				return 1;
 			}
 			char cmd[PATH_MAX];
-			if(snprintf(cmd, sizeof(cmd), "find /lib*/ -name ld-linux*>>%s 2>/dev/null", optarg) >= 0)
-				system(cmd);
-			else {
+			rc = snprintf(cmd, sizeof(cmd), "find /lib*/ -name ld-linux*>>%s 2>/dev/null", optarg);
+			assert(rc >= 0);
```
Comment (dthain) [inline parrot/src/pfs_main.cc:971]: FATAL

File: parrot/src/pfs_main.cc [hunk 34]
```diff
@@ -1210,7 +1211,8 @@ int main( int argc, char *argv[] )
 		for (int i = 0; environ[i]; i++)
 			fprintf(fp, "%s%c", environ[i], '\0');
 		char working_dir[PATH_MAX];
-		::getcwd(working_dir,sizeof(working_dir));
+		char *rc = ::getcwd(working_dir,sizeof(working_dir));
```
Comment (dthain) [inline parrot/src/pfs_main.cc:1214]: FATAL

File: parrot/src/pfs_file_cache.cc [hunk 35]
```diff
@@ -208,7 +209,10 @@ pfs_file * pfs_cache_open( pfs_name *name, int flags, mode_t mode )
 
 	fd = file_cache_open(pfs_file_cache,name->path,flags,txn,buf.st_size,0);
 	if(fd>=0) {
-		if(flags&O_TRUNC) ftruncate(fd,0);
+		if(flags&O_TRUNC) {
```
Comment (dthain) [inline parrot/src/pfs_file_cache.cc:212]: DETECT

File: dttools/src/list.c [hunk 36]
```diff
@@ -42,7 +42,8 @@ struct list_cursor {
 
 static void oom(void) {
 	const char *message = "out of memory\n";
-	write(STDERR_FILENO, message, strlen(message));
+	ssize_t rc = write(STDERR_FILENO, message, strlen(message));
```
Comment (dthain) [inline dttools/src/list.c:45]: IGNORE

File: makeflow/src/dag_visitors.c [hunk 37]
```diff
@@ -355,7 +356,8 @@ void dag_to_dax_replica_catalog(const struct dag *d, FILE *output)
 	list_first_item(input_files);
 	while((f = (struct dag_file*)list_next_item(input_files)))
 	{
-		realpath(f->filename, fn);
+		char *rc = realpath(f->filename, fn);
```
Comment (dthain) [inline makeflow/src/dag_visitors.c:359]: FATAL

File: makeflow/src/dag_visitors.c [hunk 38]
```diff
@@ -383,7 +385,8 @@ void dag_to_dax_transform_catalog(const struct dag *d, FILE *output)
 	list_first_item(transforms);
 	while((fn = list_next_item(transforms))) {
 		if(path_lookup(getenv("PATH"), fn, pfn, PATH_MAX)){
-			realpath(fn, pfn);
+			char *rc = realpath(fn, pfn);
+			assert(rc != NULL);
```
Comment (dthain) [inline makeflow/src/dag_visitors.c:389]: FATAL

File: makeflow/src/makeflow.c [hunk 39]
```diff
@@ -1125,7 +1125,8 @@ static void handle_abort(int sig)
 	if (fd >= 0) {
 		char buf[256];
 		string_nformat(buf, sizeof(buf), "received signal %d (%s), cleaning up remote jobs and files...\n",sig,strsignal(sig));
-		write(fd, buf, strlen(buf));
+		ssize_t rc = write(fd, buf, strlen(buf));
```
Comment (dthain) [inline makeflow/src/makeflow.c:1128]: IGNORE

File: makeflow/src/makeflow.c [hunk 40]
```diff
@@ -2336,8 +2337,10 @@ int main(int argc, char *argv[])
 
 	makeflow_parse_input_outputs(d);
 
-	if (change_dir)
-		chdir(change_dir);
+	if (change_dir) {
+		int rc = chdir(change_dir);
```
Comment (dthain) [inline makeflow/src/makeflow.c:2341]: FATAL

File: makeflow/src/makeflow_analyze.c [hunk 41]
```diff
@@ -327,7 +328,8 @@ int main(int argc, char *argv[])
 		char expanded_path[PATH_MAX];
 
 		collect_input_files(d, bundle_directory, bundler_rename);
-		realpath(bundle_directory, expanded_path);
+		char *rc = realpath(bundle_directory, expanded_path);
```
Comment (dthain) [inline makeflow/src/makeflow_analyze.c:331]: FATAL

File: makeflow/src/makeflow_gc.c [hunk 42]
```diff
@@ -202,7 +202,8 @@ void makeflow_clean_node(struct dag *d, struct batch_queue *queue, struct dag_no
 		char *command = string_format("%s --clean",n->task->command);
 		printf("%s\n",command);
 		jx_export(n->task->envlist);
-		system(command);
+		int rc = system(command);
```
Comment (dthain) [inline makeflow/src/makeflow_gc.c:205]: DETECT

File: makeflow/src/makeflow_linker.c [hunk 43]
```diff
@@ -115,7 +116,8 @@ file_type file_unix_file_known(const char *name){
 
 	FILE *unix_file_pipe = popen(file, "r");
 	char *unix_file_output = (char *) malloc(1024 * sizeof(char));
-	fgets(unix_file_output, 1024*sizeof(char), unix_file_pipe);
+	char *rc = fgets(unix_file_output, 1024*sizeof(char), unix_file_pipe);
```
Comment (dthain) [inline makeflow/src/makeflow_linker.c:119]: DETECT

File: makeflow/src/makeflow_linker.c [hunk 44]
```diff
@@ -143,7 +145,8 @@ void initialize( char *output_directory, char *input_file, struct list *d){
 	create_workspace();
 
 	char expanded_input[PATH_MAX];
-	realpath(input_file, expanded_input);
+	char *rc = realpath(input_file, expanded_input);
+	assert(rc != NULL);
```
Comment (dthain) [inline makeflow/src/makeflow_linker.c:149]: FATAL

File: makeflow/src/makeflow_linker.c [hunk 45]
```diff
@@ -158,7 +161,8 @@ void initialize( char *output_directory, char *input_file, struct list *d){
 struct list *find_dependencies_for(struct dependency *dep){
 	pid_t pid;
 	int pipefd[2];
-	pipe(pipefd);
+	int rc = pipe(pipefd);
+	assert(rc == 0);
```
Comment (dthain) [inline makeflow/src/makeflow_linker.c:165]: DETECT

File: makeflow/src/makeflow_linker.c [hunk 46]
```diff
@@ -477,14 +481,16 @@ int main(int argc, char *argv[]){
 	initialize(output, input, dependencies);
 	char *tmp = output;
 	output = (char *) malloc(PATH_MAX * sizeof(char));
-	realpath(tmp, output);
+	char *rc = realpath(tmp, output);
+	assert(rc != NULL);
```
Comment (dthain) [inline makeflow/src/makeflow_linker.c:485]: FATAL

File: makeflow/src/makeflow_linker.c [hunk 47]
```diff
@@ -477,14 +481,16 @@ int main(int argc, char *argv[]){
...
 	free(tmp);
 	if(!create_dir(output, 0777)) fatal("Could not create output directory.\n");
 	if(verbose) fprintf(stdout, "Using %s as output location.\n", output);
 
 	char input_wd[PATH_MAX];
 	path_dirname(input, input_wd);
-	chdir(input_wd);
+	int rc2 = chdir(input_wd);
+	assert(rc2 == 0);
```
Comment (dthain) [inline makeflow/src/makeflow_linker.c:493]: FATAL

File: parrot/src/parrot_package_create.c [hunk 48]
```diff
@@ -262,7 +262,8 @@ int prepare_work()
 		}
 		char mkdir_cmd[PATH_MAX * 2];
 		if(snprintf(mkdir_cmd, sizeof(mkdir_cmd), "mkdir -p %s", packagepath) >= 0) {
-			system(mkdir_cmd);
+			int rc = system(mkdir_cmd);
```
Comment (dthain) [inline parrot/src/parrot_package_create.c:265]: DETECT

File: parrot/src/parrot_package_create.c [hunk 49]
```diff
@@ -825,9 +826,9 @@ int main(int argc, char *argv[])
 	char special_filename_tmp[PATH_MAX];
 	string_nformat(special_filename_tmp, sizeof(special_filename_tmp), "%s%s", special_filename, ".tmp");
 	char sort_cmd[PATH_MAX * 2];
-	if(snprintf(sort_cmd, PATH_MAX * 2, "sort -u %s>>%s", special_filename, special_filename_tmp) >= 0)
-		system(sort_cmd);
-	else {
+	int rc = snprintf(sort_cmd, PATH_MAX * 2, "sort -u %s>>%s", special_filename, special_filename_tmp);
```
Comment (dthain) [inline parrot/src/parrot_package_create.c:829]: FATAL

File: parrot/src/pfs_channel.c [hunk 50]
```diff
@@ -108,7 +109,8 @@ int pfs_channel_init( pfs_size_t size )
 	}
 
 	channel_size = size;
-	ftruncate(channel_fd,channel_size);
+	int rc = ftruncate(channel_fd,channel_size);
```
Comment (dthain) [inline parrot/src/pfs_channel.c:112]: DETECT

File: parrot/src/pfs_channel.c [hunk 51]
```diff
@@ -193,7 +195,8 @@ int pfs_channel_alloc( const char *name, pfs_size_t length, pfs_size_t *start )
 			debug(D_CHANNEL,"channel expanded to 0x%" PRIu64 " bytes at base 0x%" PRIdPTR, newsize, (uintptr_t) newbase);
 			return pfs_channel_alloc(name,length,start);
 		}
-		ftruncate64(channel_fd,channel_size);
+		int rc = ftruncate64(channel_fd,channel_size);
```
Comment (dthain) [inline parrot/src/pfs_channel.c:198]: DETECT

**Comment by dthain** [comment]
An adjustment to my prior comment:

For many of the FATAL cases, I think we should introduce several functions `xxgetcwd` `xxchdir` `xxrealpath` that either succeed or call `fatal` and return `void`.  Of course, they shouldn't be used everywhere, but only in cases where there is no point in further progress if, for example, the working directory does not exist.

**Pull request closed without merging**